pkg/controller/job/job_controller.go

```diff
@@ -487,8 +487,12 @@ func (jm *JobController) syncJob(key string) (bool, error) {
 
 	jobHaveNewFailure := failed > job.Status.Failed
 
-	// check if the number of failed jobs increased since the last syncJob
-	if jobHaveNewFailure && (int32(previousRetry)+1 > *job.Spec.BackoffLimit) {
+	if job.Spec.Template.Spec.RestartPolicy == v1.RestartPolicyOnFailure && pastBackoffLimit(&job, pods) {
+		jobFailed = true
+		failureReason = "BackoffLimitExceeded"
+		failureMessage = "Job has reach the specified backoff limit"
+	} else if jobHaveNewFailure && (int32(previousRetry)+1 > *job.Spec.BackoffLimit) {
```

Comment:
Combine the first two if-cases

Reply:
I intentionally put them separately to avoid having monstrous if condition, but since you ask ;-)